### PR TITLE
[PATCH v4] test: ipsecfwd: add option for direct reception 

### PR DESCRIPTION
**Pull request by TuomasTaipale**
This patchset adds an option to not use scheduler with `odp_ipsecfwd`. Tester throughput increases slightly as scheduling overhead disappears. 

Additionally, packets are now forwarded with bursts instead of single packet at a time after lookup, this further increases tester thoughput if a lot of packets are destined to the same output per processing loop.

v2:
- Added a separate patch where existing processing functions are moved earlier
- Changed "no scheduler" option to "direct RX" option and changed title accordingly
- Rebased

v3: 
- Addressed Janne's comments

v4:
- Added reviewed-by tag

### CODE REVIEW

File: test/performance/odp_ipsecfwd.c [hunk 1]
```diff
@@ -329,6 +345,9 @@ static void print_usage(void)
 	       "  -I, --num_input_qs  Input queue count. 1 by default.\n"
 	       "  -S, --num_sa_qs     SA queue count. 1 by default.\n"
 	       "  -O, --num_output_qs Output queue count. 1 by default.\n"
+	       "  -d, --direct_rx     Use direct RX. Interfaces will be polled by workers\n"
+	       "                      directly. \"--mode\" and \"--num_input_qs\" options are\n"
+	       "                      ignored, input queue count will match worker count.\n"
```
Comment (JannePeltonen) [inline test/performance/odp_ipsecfwd.c:358]: Now it would be quite straightforward to support sync ipsec also with scheduled I/O, but I suppose that could be a topic for another patch set. Or if you want to add it now, I would suggest adding sync mode first and then the direct mode I/O on top if it in a separate patch.

File: test/performance/odp_ipsecfwd.c [hunk 2]
```diff
@@ -622,28 +637,224 @@ static void drain_events(void)
...
+static inline int process_ipsec_in(odp_packet_t pkts[], const odp_ipsec_sa_t sas[], int num,
+				   odp_packet_t pkts_out[])
+{
+	odp_ipsec_in_param_t param;
+	int left, sent = 0, num_out, ret;
+
+	memset(&param, 0, sizeof(param));
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	for (int i = 0; i < num; ++i)
+		odp_packet_user_ptr_set(pkts[i], NULL);
```
Comment (JannePeltonen) [inline test/performance/odp_ipsecfwd.c:675]: This is not needed in the sync mode here nor in process_ipsec_out().

File: test/performance/odp_ipsecfwd.c [hunk 3]
```diff
@@ -489,12 +501,26 @@ static inline uint32_t forward_packets(odp_packet_t pkts[], int num, odph_table_
 			odp_packet_free(pkt);
 			continue;
 		}
-		if (odp_unlikely(!pktio->send_fn(pktio, hash % pktio->num_tx_qs, pkt))) {
-			odp_packet_free(pkt);
+
+		vec = &vecs[pktio->idx];
+		vec->pkts[vec->num++] = pkt;
+		vec->hash = hash;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsecfwd.c:507]: This effectively overwrites the flow hash of the preceding packets in the vector, causing them possibly go out through the wrong output queue, which in turn can break packet order in the ordered scheduled mode.

File: test/performance/odp_ipsecfwd.c [hunk 4]
```diff
@@ -489,12 +501,26 @@ static inline uint32_t forward_packets(odp_packet_t pkts[], int num, odph_table_
...
+
+	for (uint32_t i = 0U; i < MAX_IFS; ++i) {
+		if (vecs[i].num == 0U)
 			continue;
-		}
 
-		++num_procd;
+		vec = &vecs[i];
+		pktio = vec->pktio;
+		uint32_t ret = pktio->send_fn(pktio, vec->hash % pktio->num_tx_qs, vec->pkts,
```
Comment (JannePeltonen) [inline test/performance/odp_ipsecfwd.c:517]: Hash based queue selection is not needed in direct mode (and is not needed in the unoredered scheduled mode that was already supported) since packet order across threads is not maintained anyway. Thread-specific output queues would be ok and fast (and could be used in the MT_UNSAFE mode) with direct pktio output.

File: test/performance/odp_ipsecfwd.c [hunk 5]
```diff
@@ -1365,8 +1401,6 @@ static odp_bool_t setup_pktios(prog_config_t *config)
 		pktio->num_tx_qs = config->num_output_qs;
 		odp_pktout_queue_param_init(&pktout_param);
 		pktout_param.num_queues = pktio->num_tx_qs;
-		pktout_param.op_mode = config->num_thrs > (int)pktio->num_tx_qs ?
-			ODP_PKTIO_OP_MT : ODP_PKTIO_OP_MT_UNSAFE;
```
Comment (JannePeltonen) [inline test/performance/odp_ipsecfwd.c]: This looks like a bug fix that probably should be in its own patch before the other patches of this PR.

File: test/performance/odp_ipsecfwd.c (hunk 1 above)
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: Leaving this as future work.

File: test/performance/odp_ipsecfwd.c (hunk 2 above)
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: Ditched in v3.

File: test/performance/odp_ipsecfwd.c (hunk 3 above)
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: Overhauled this in v3.

File: test/performance/odp_ipsecfwd.c (hunk 4 above)
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: New patch (06ba5ae) for this in v3.

File: test/performance/odp_ipsecfwd.c [hunk 6]
```diff
@@ -1020,7 +1046,7 @@ static void parse_sas(prog_config_t *config)
 	(void)fclose(file);
 }
 
-static const pktio_t *get_pktio(const char *iface, const prog_config_t *config)
+static const pktio_t *get_pktio(const char *iface, prog_config_t *config)
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: Me neither, unintentional, fixed in v3.

File: test/performance/odp_ipsecfwd.c (hunk 5 above)
Comment (TuomasTaipale) [reply test/performance/odp_ipsecfwd.c]: Moved to its own patch (6865d09) in v3.
